api/submission/remove_submission_test.go

```diff
@@ -0,0 +1,71 @@
+package submission
+
+import (
+    "testing"
+    
+    "github.com/eriq-augustine/autograder/api/core"
+    "github.com/eriq-augustine/autograder/db"
+    "github.com/eriq-augustine/autograder/model"
+)
+
+func TestRemoveSubmission(test *testing.T) {
+    // Leave the course in a good state after the test.
+    defer db.ResetForTesting();
+
+    testCases := []struct{ role model.UserRole; targetEmail string; targetSubmission string; foundUser bool; foundSubmission bool; permError bool}{
+        // Grader, self, recent.
+        {model.RoleGrader, "grader@test.com", "", true, false, false},
+        {model.RoleGrader, "grader@test.com", "", true, false, false},
+
+        // Grader, self, missing.
+        {model.RoleGrader, "",                "ZZZ", true, false, false},
+        {model.RoleGrader, "grader@test.com", "ZZZ", true, false, false},
+
+        // Grader, other, recent.
+        {model.RoleGrader, "student@test.com", "",true, true, false},
+
+        // Grader, other, specific.
+        {model.RoleGrader, "student@test.com", "1697406256", true, true, false, },
+        {model.RoleGrader, "student@test.com", "1697406265", true, true, false},
+        {model.RoleGrader, "student@test.com", "1697406272", true, true, false},
+
+        // Grader, other, specific (full ID).
+        {model.RoleGrader, "student@test.com", "course101::hw0::student@test.com::1697406256", true, true, false},
+        {model.RoleGrader, "student@test.com", "course101::hw0::student@test.com::1697406265", true, true, false},
+        {model.RoleGrader, "student@test.com", "course101::hw0::student@test.com::1697406272", true, true, false},
+
+        // Grader, other, missing.
+        {model.RoleGrader, "student@test.com", "ZZZ", true, false, true},
+
+        // Grader, missing, recent.
+        {model.RoleGrader, "ZZZ@test.com", "", false, false, true},
+    };
```

Comment:
Add in a set of cases that sweep through the other roles (users),
making sure all below grader get a perm (permissions) error.

Reply:
(Unresolving)

This has not been resolved.
Look at a test like api/user/remove:
https://github.com/eriq-augustine/autograder-server/blob/main/api/user/remove_test.go#L23

All (but one) of the student tests you added are redundant, since they will all perm error.

Reply:
If the grader role can pass all tests, can we assume admin and owner will too or does that also require a set of test cases, and if so, how thorough should the tests be?

Reply:
Your assumption is not wrong, but why rely on an assumption when it is trivial to test?
(Only five copy-paste lines.)

This is actually a really important lesson to learn about when to just trust assumptions and when to test things.
Like I mentioned above, in this case the cost of testing is so low compared to the safety it buys is that there is no reason not to do it.

Reply:
Ok so as a follow up, there's 13 grader-role tests that test all scenarios. To be extra safe, wouldn't it make sense to use the same 13 test cases for Admin and Owner then? Also I'm a bit confused on the unknown role as it doesn't return a bad permission error (-306), but it is lower than a grader.

Reply:
No, not necessary.
There is no functionality difference between any of the allowed roles (it's the same for an admin and owner).
So, we just need to make sure that those roles are not denied.
We typically choose to run tests at the lowest allowed permission level, but aside from that the choice of primary user is not too important.

Regarding the unknown role.
No user should have this role, so it will probably be caught even earlier in the system than a permissions error.
If no error is returned, then it is a problem.
An unknown role is like a nil role.